src/main/java/sparksoniq/jsoniq/runtime/iterator/operational/TypePromotionIterator.java

```diff
@@ -0,0 +1,173 @@
+    }
+
+    private void setNextResult() {
+        _nextResult = null;
+        if (_iterator.hasNext()) {
+            _nextResult = _iterator.next();
+            if (_nextResult != null)
+                _childIndex++;
+        } else {
+            _iterator.close();
+            checkEmptySequence(_childIndex);
+        }
+
+        this._hasNext = _nextResult != null;
+        if (!hasNext())
+            return;
+
+        checkItemsSize(_childIndex);
+        if (!_nextResult.isTypeOf(itemType)) {
+            checkTypePromotion();
+        }
+    }
+
+    private void checkEmptySequence(long size) {
+        if (
+            size == 0
+                && (_sequenceType.getArity() == SequenceType.Arity.One
+                    ||
+                    _sequenceType.getArity() == SequenceType.Arity.OneOrMore)
+        ) {
+            throw new UnexpectedTypeException(
+                    "Empty sequence cannot be promoted to type "
+                        + sequenceTypeName
+                        + _sequenceType.getArity().getSymbol(),
+                    getMetadata()
+            );
+        }
+    }
+
+    private void checkItemsSize(long size) {
+        if (size > 0 && _sequenceType.isEmptySequence()) {
+            throw new UnexpectedTypeException(
+                    ItemTypes.getItemTypeName(_nextResult.getClass().getSimpleName())
+                        + " cannot be promoted to type empty-sequence()",
+                    getMetadata()
+            );
+        }
+
+
+        if (
+            size > 1
+                && (_sequenceType.getArity() == SequenceType.Arity.One
+                    ||
+                    _sequenceType.getArity() == SequenceType.Arity.OneOrZero)
+        ) {
+            throw new UnexpectedTypeException(
+                    "Sequences of more than one item cannot be promoted to type "
+                        + sequenceTypeName
+                        + _sequenceType.getArity().getSymbol(),
+                    getMetadata()
+            );
+        }
+    }
+
+    @Override
+    public boolean initIsRDD() {
+        return _iterator.isRDD();
+    }
+
+    @Override
+    public JavaRDD<Item> getRDDAux(DynamicContext context) {
+        _currentDynamicContext = context;
+        JavaRDD<Item> childRDD = _iterator.getRDD(context);
+
+        int count = childRDD.take(2).size();
+        checkEmptySequence(count);
+        checkItemsSize(count);
+        Function<Item, Item> transformation = new TypePromotionClosure(_exceptionMessage, _sequenceType, getMetadata());
+        return childRDD.map(transformation);
+    }
+
+    private void checkTypePromotion() {
+        if (!_nextResult.canBePromotedTo(_sequenceType.getItemType()))
+            throw new UnexpectedTypeException(
+                    _exceptionMessage
+                        + ItemTypes.getItemTypeName(_nextResult.getClass().getSimpleName())
+                        + " cannot be promoted to type "
```

Comment:
This message is correct -- this is an atomic value that cannot be promoted to a type.